### ISSUE TITLE
cassandra microservice app fails to start with docker-compose generated conf 

### ISSUE THREAD

**Issue by raphaelbrugier** [jhipster/generator-jhipster#3427]
##### **Overview of the issue**
When adding a microservice app with a Cassandra database, the app fails to start when launched with the configuration generated by the docker-compose generator.
The reason is the Cassandra cluster is not yet initialized (no keyspace or tables created) when the app starts. The setup of the cluster has to be done manually.

##### **JHipster Version(s)** 
master

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder** 

3 directories: 
- jhigate
- casstest (cassandra microservice)
- compose (docker-compose config)

**gateway:**
```
{
  "generator-jhipster": {
    "jhipsterVersion": "3.0.0",
    "baseName": "jhigate",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "hibernateCache": "hazelcast",
    "clusteredHttpSession": "no",
    "websocket": "no",
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": "no",
    "buildTool": "maven",
    "jwtSecretKey": "080bce90a1e6565afd977f09ce38b41dc315b080",
    "useSass": false,
    "applicationType": "gateway",
    "testFrameworks": [
      "gatling"
    ],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en"
    ]
  }
}
```

**cassandra microservice:**
```
{
  "generator-jhipster": {
    "jhipsterVersion": "3.0.0",
    "baseName": "casstest",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "serverPort": "8085",
    "authenticationType": "jwt",
    "hibernateCache": "no",
    "databaseType": "cassandra",
    "devDatabaseType": "cassandra",
    "prodDatabaseType": "cassandra",
    "searchEngine": "no",
    "buildTool": "maven",
    "jwtSecretKey": "a310878c074bd02c67adc51c484c420978c039aa",
    "enableTranslation": false,
    "applicationType": "microservice",
    "testFrameworks": [
      "gatling"
    ],
    "jhiPrefix": "jhi",
    "skipClient": true,
    "skipUserManagement": true,
    "clusteredHttpSession": "no",
    "websocket": "no",
    "enableSocialSignIn": false
  }
}
```

**compose:**
```
{
  "generator-jhipster": {
    "appsFolders": [
      "casstest",
      "jhigate"
    ],
    "directoryPath": "../",
    "useElk": true,
    "jwtSecretKey": "1523fb4eff5ba4c0966a8c69a10bd136e0a9d37d"
  }
}
```

##### **Browsers and Operating System** 
OS/X

##### **Reproduce the error** 
Run `yo jhispter` in each app subdirectory
Run `./mvnw package -Pprod -DskipTests docker:build` in the gateway and casstest directories
Run `yo jhipster:docker-compose` in compose directory
Run `docker-compose up -d` in compose directory to start the cluster and all containers

##### **Related issues** 
#3407
#3288

##### **Suggest a Fix** 
A manual workaround is to expose the access to the Cassandra cluster (port 9042),  then connect with cqlsh then run the queries manually.

An automatic fix would be to reuse the container and scripts introduced in #3288 to automatically setup the Cassandra cluster before starting the app.

We need to discuss this, depending on the intent behind the docker-compose generated files.
- If it's only intended for development test, it could be ok.
- If it's also intended to be use in production, it could be dangerous to automatically run migration queries. (Although it's what it is done with SQL databases I think.)

An option could be to ask the user if he wants automatic Cassandra migrations when generated the file.

**Comment by pascalgrimaud** [comment]
@raphaelbrugier 
What do you think about managing our own table changelog (like mongeez, mongobee):
- create table "changelog"
- insert the name of the cql script when it has been executed successfully
- ignore the execution when the cql script exists in the table "changelog"

**Comment by raphaelbrugier** [comment]
Yes, this would be the second step of the automatic migration feature.
Having something similar to liquibase / mongeez
I'll propose a PR for this.

My question however was more on the intent of the docker-compose configuration for production, not only for Cassandra but also for the other databases.
Is it common to let the application execute the migration queries automatically when deployed in production?

**Comment by pascalgrimaud** [comment]
In my opinion, all docker-compose files generated by JHipster are intended for development.
`cassandra-cluster.yml` and `cassandra-opscenter.yml` are here to provide an architecture which could be similar as in production.

In a real project, I think only the **package.war** is given to production.
If there is an infrastructure with Docker, DevOps should work to provide all Dockerfiles, similar for Prod and for Dev. But, as I don't have experience with a customer who uses Docker in Prod, I don't know :)

**Comment by raphaelbrugier** [comment]
Oh ok, then there is a confusion about when to use the generated docker-compose files.
I assumed it was intended for both development and production, especially because in the [documentation](http://jhipster.github.io/docker-compose/) it is described as:

> Using Docker and Docker Compose is highly recommended in development, and is also a good solution in production.

I've met some people using docker in production here in DC, not with a JHipster stack however.

Also, even when not using docker-compose in production, the changesets automatically applied by the application itself right?
I'm thinking about this [class](https://github.com/jhipster/jhipster-sample-app-token/blob/master/src/main/java/com/mycompany/myapp/config/liquibase/AsyncSpringLiquibase.java) applying the liquibase algorithm when the application is started, synchronously in production.

**Comment by pascalgrimaud** [comment]
After merging https://github.com/jhipster/generator-jhipster/pull/3544, which is a very good step, I have some comments:
1) supposed to be used for dev, should we keep this part ? https://github.com/jhipster/generator-jhipster/blob/master/generators/server/templates/src/main/docker/_cassandra.yml#L14-L19
2) it's just my opinion, it would be interesting to use this new system with `schema_version` for **cassandra-cluster** as well => https://github.com/jhipster/generator-jhipster/blob/master/generators/server/templates/src/main/docker/_cassandra.yml#L20-L32
3) +1 for moving `create-tables.cql`

**Comment by raphaelbrugier** [comment]
1) even for dev, I would say it's ok to leave a second node in the Cluster.
all C* cluster always run with more than one node, so let's keep this in dev. the more you dev environment is close to production, the better!

2) Ok, this will solve the initial reported problem

3) Ok, I'll work on that. (I've asked you on the other PR before seeing you +1 here)

**Comment by raphaelbrugier** [comment]
I'm closing this as it was fixed by #3593
